docs/design/2020-09-28-cluster-read-only-mode.md

```diff
@@ -0,0 +1,233 @@
+# Proposal: Support Global Read Only Mode
+
+- Author(s):    [ichn-hu](https://github.com/ichn-hu)
+- Last updated:  Oct. 09, 2020
+- Discussion at:  
+
+## Abstract
+
+This document describes the design of the `read_only` global variable that once turned on blocks non-super users from updating the TiDB cluster.
+
+
+## Motivation or Background
+
+As requested and described in [this issue](https://github.com/pingcap/tidb/issues/18426), such a variable can be beneficial to maintenance, migration and replication, and according to a [blog post](https://pingcap.com/blog/heterogeneous-database-replication-to-tidb#creating-a-read-only-or-archive-database), TiDB can be used as a read-only or archive database, in which case we would like to make the cluster running in a read only mode, while only allowing CDC tools to update the data.
+
+
+## Detailed Design
+
+Based on the issue description and [MySQL’s description](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_read_only), the `read_only` option of TiDB will act similarly.
+
+
+### Behavior Description
+
+
+*   `read_only` is disabled by default.
+*   The only way to enable `read_only` mode is to execute `set global read_only = on` with a SuperPriv user.
+    *   The enabling attempt will block until the whole cluster enters the `read_only` mode.
+    *   And only `set global read_only = off` to turn off the mode.
+    *   The `read_only` global variable will have 3 values, the reason for such design is to make the status more intuitive, see section technical issues 4 for a detailed explanation.
+        *   off, when the cluster acts in a normal mode
+        *   enabling, when the turning on process is going on
+        *   on, when the whole cluster is in read only mode
+    *   `set global read_only = on` upon return
+        *   if there are warnings, meaning there are still some TiDB servers in the cluster that are possibly not read only, the user should check if these TiDB servers are down or have unreleased write lock or have large transactions committing.
+            *   The warnings contain all known TiDB servers that have not acknowledged the read only mode.
+            *   If the user examines the `read_only` global variable, it will be `enabling`, not `on`.
+            *   She could rerun `set global read_only = on` to wait for another period of time (default to 10 seconds, could have exponential backoff behavior).
```

Comment:
> I think we should clarify the language here. If changing to `on` fails, an error should be returned and not a warning. This is the typical behavior when a type of lock fails. Because the lock failed, I also expect that the read_only variable is changed from `ENABLING` back to `OFF.

It is not `fail`, it is just the enabling attempt takes too long because some TiDB nodes have not responded, and some server are already `read_only`(they will reject queries that might update data).

Normally the attempt should work out without timeout, unless there are idle TiDB that are not responding, and we need user intervention in this case.

Also note the user could run `set global read_only = off` at any time they want to cancel the enabling. If we explicitly tell the users that they should not assume the cluster is read only unless its value is `on`, I think it is acceptable.